src/cdomains/arrayDomain.ml

```diff
@@ -567,8 +567,21 @@ struct
   let smart_widen = smart_widen_with_length None
   let smart_leq = smart_leq_with_length None
 
-  let meet a b = normalize @@ meet a b
-  let narrow a b = normalize @@ narrow a b
+  let meet (e1,v1) (e2,v2) = normalize @@
+    match e1,e2 with
+    | `Lifted e1e, `Lifted e2e when not (Basetype.CilExp.equal e1e e2e) ->
+      (* partitioned according to two different expressions -> meet can not be element-wise *)
+      (* TODO: do smart things if the relationship between e1e and e2e is known *)
+      (e1,v1)
```

Comment:
I guess it's safe to return the first one because if it's partitioned by multiple expressions, their meet is partitioned by both, but we can only keep one, so arbitrarily prefer the first?

Seems weird that it's not commutative though, but I have no idea what else could be done here. Maybe worth a code comment pointing this out?

Reply:
I added a comment to this effect. In `join` there is some elaborate logic on what expression to keep. In #485, that logic should be applied everywhere (making a `smart_meet` and `smart_narrow` necessary).